### DISCUSSION TITLE
How can I check the information of recipents and senders in misigAprrove message?

### DISCUSSION THREAD

**Discussion by jinho-korbit** [filecoin-project/lotus#9105] [General Q&A] [answered]
How can I check the information of recipents and senders in misigAprrove message?

I need information on the details of the remittance from multisig wallet, so I want to decode the encoded data
to check this part, but there is a problem that it can't be decode. 

I wonder what to do in this case.
Please refer to the attached API details that I call.

message ID : bafy2bzacebv27y3afblxarmw5uxohajt7zr6mrmjkg2vy22atlvns2wosf3rm

1. First Call Get Message Detail API Request Message and Response Message
curl -X POST 'https://api.node.glif.io' -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":1,"method":"Filecoin.ChainGetMessage","params":[
    {
        "/": "bafy2bzacebv27y3afblxarmw5uxohajt7zr6mrmjkg2vy22atlvns2wosf3rm"
    }
]}'

ChainGetMessage result
{
    "jsonrpc": "2.0",
    "result": {
        "Version": 0,
        "To": "f2kemhdxzy6lc2zt2c7gzr6h5d3mlurcbllfodqyq",
        "From": "f1lkmgrfr6o6n3nzqtidaogsmegqjfhnlbupiixuy",
        "Nonce": 56,
        "Value": "0",
        "GasLimit": 8397753,
        "GasFeeCap": "773391486",
        "GasPremium": "337364",
        "Method": 3,
        "Params": "ghibQA==",
        "CID": {
            "/": "bafy2bzacebehkylyflxlmwoszxl235hxly4vremnkgzcnd3d6hxygmvnl6mc4"
        }
    },
    "id": 1
}

2. Call Filecoin.StateCall API
- Request Message
curl -X POST 'https://api.node.glif.io' -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":1,"method":"Filecoin.StateCall","params":[
    {
        "Version": 0,
        "To": "f2kemhdxzy6lc2zt2c7gzr6h5d3mlurcbllfodqyq",
        "From": "f1lkmgrfr6o6n3nzqtidaogsmegqjfhnlbupiixuy",
        "Nonce": 56,
        "Value": "0",
        "GasLimit": 8397753,
        "GasFeeCap": "773391486",
        "GasPremium": "337364",
        "Method": 3,
        "Params": "ghibQA==",
        "CID": {
            "/": "bafy2bzacebehkylyflxlmwoszxl235hxly4vremnkgzcnd3d6hxygmvnl6mc4"
        }
    },
    [
        {"/":"bafy2bzacean2oxhirt3utlic6shad366sbft74fkv5d3jvy7onh5i77kto23e"},{"/":"bafy2bzaceaijr62tvtfsvv6h33a4tf2d7ulfrre7bjbtpfx7f7wpniqwldohg"},{"/":"bafy2bzaceca6znbfvemp4fygq3cj2yegutqafsizjzp2zady5qg5qz7he646u"}
    ]
]}'

- Response Message
{
    "jsonrpc": "2.0",
    "result": {
        "MsgCid": {
            "/": "bafy2bzacebehkylyflxlmwoszxl235hxly4vremnkgzcnd3d6hxygmvnl6mc4"
        },
        "Msg": {
            "Version": 0,
            "To": "f2kemhdxzy6lc2zt2c7gzr6h5d3mlurcbllfodqyq",
            "From": "f1lkmgrfr6o6n3nzqtidaogsmegqjfhnlbupiixuy",
            "Nonce": 56,
            "Value": "0",
            "GasLimit": 8397753,
            "GasFeeCap": "773391486",
            "GasPremium": "337364",
            "Method": 3,
            "Params": "ghibQA==",
            "CID": {
                "/": "bafy2bzacebehkylyflxlmwoszxl235hxly4vremnkgzcnd3d6hxygmvnl6mc4"
            }
        },
        "MsgRct": {
            "ExitCode": 0,
            "Return": "g/UAQA==",
            "GasUsed": 6451840
        },
        "GasCost": {
            "Message": null,
            "GasUsed": "0",
            "BaseFeeBurn": "0",
            "OverEstimationBurn": "0",
            "MinerPenalty": "0",
            "MinerTip": "0",
            "Refund": "0",
            "TotalCost": "0"
        },
        "ExecutionTrace": {
            "Msg": {
                "Version": 0,
                "To": "f2kemhdxzy6lc2zt2c7gzr6h5d3mlurcbllfodqyq",
                "From": "f1lkmgrfr6o6n3nzqtidaogsmegqjfhnlbupiixuy",
                "Nonce": 56,
                "Value": "0",
                "GasLimit": 8397753,
                "GasFeeCap": "773391486",
                "GasPremium": "337364",
                "Method": 3,
                "Params": "ghibQA==",
                "CID": {
                    "/": "bafy2bzacebehkylyflxlmwoszxl235hxly4vremnkgzcnd3d6hxygmvnl6mc4"
                }
            },
            "MsgRct": {
                "ExitCode": 0,
                "Return": "g/UAQA==",
                "GasUsed": 6451840
            },
            "Error": "",
            "Duration": 4818562,
            "GasCharges": null,
            "Subcalls": null
        },
        "Error": "",
        "Duration": 4818562
    },
    "id": 1
}

3. Call StateDecodeParams API for decoding encoded data in a Statecall response message 
curl -X POST 'https://api.node.glif.io' -H 'Content-Type: application/json' --data '{"jsonrpc":"2.0","id":1,"method":"Filecoin.StateDecodeParams","params":[
    "f01554291",
    3,
    "ghibQA==",
    [
        {"/":"bafy2bzacean2oxhirt3utlic6shad366sbft74fkv5d3jvy7onh5i77kto23e"},{"/":"bafy2bzaceaijr62tvtfsvv6h33a4tf2d7ulfrre7bjbtpfx7f7wpniqwldohg"},{"/":"bafy2bzaceca6znbfvemp4fygq3cj2yegutqafsizjzp2zady5qg5qz7he646u"}
    ] 
]}'

but.... Error occur
{"jsonrpc":"2.0","id":1,"error":{"code":1,"message":"getting params type: unknown method 3 for actor bafk2bzacedudbf7fc5va57t3tmo63snmt3en4iaidv4vo3qlyacbxaa6hlx6y"}}

Help me...
I want to know how to decode the params and how to decode the return value of MsgRct.
"MsgRct": {
       "ExitCode": 0,
       "Return": "g/UAQA==",
        "GasUsed": 6451840
}

**Comment by LexLuthr** [comment, marked as answer]
Your params are incorrect here in the last API call.

// StateDecodeParams attempts to decode the provided params, based on the recipient actor address and method number.
	StateDecodeParams(ctx context.Context, toAddr address.Address, method abi.MethodNum, params []byte, tsk types.TipSetKey) (interface{}, error) //perm:read
	
	
You need to provide `toAddr address.Address` which would be `f2kemhdxzy6lc2zt2c7gzr6h5d3mlurcbllfodqyq` in your case.

```
# lotus chain decode params f2kemhdxzy6lc2zt2c7gzr6h5d3mlurcbllfodqyq 3 "ghibQA=="
2022/08/01 07:03:57 proto: duplicate proto type registered: pb.NoiseHandshakePayload
{
  "ID": 155,
  "ProposalHash": null
}

```

**Reply by jinho-korbit** [reply]
very very Thank you!!!!!

**Reply by jinho-korbit** [reply]
@LexLuthr 

how to  decode below encoding data?

"MsgRct": {
"ExitCode": 0,
"Return": "g/UAQA==",
"GasUsed": 6451840
}

**Reply by LexLuthr** [reply]
That's a little bit more tricky. I don't think there is any direct command for this. 
You will need to read these byte and supply it to appropriate return's UnmarshalCBOR. You will need to know before hand what type of return are you expecting from the message.

For example this is a Multisig Approve Message:

```
var ret multisig.ApproveReturn
		 ret.UnmarshalCBOR(bytes.NewReader(wait.Receipt.Return))
```

**Reply by jinho-korbit** [reply]
@LexLuthr 
So, how do I get the following information from multiApprove meesage?

"transfers": [
{
"from": "f1uhvv3dmvveaycnedz4pl43t6ptsvsdu2acdzniq",
"fromId": "f01816145",
"to": "f0131822",
"toId": "f0131822",
"value": "12699996651843",
"type": "miner-fee"
},
{
"from": "f1uhvv3dmvveaycnedz4pl43t6ptsvsdu2acdzniq",
"fromId": "f01816145",
"to": "f099",
"toId": "f099",
"value": "25540157027699286",
"type": "burn-fee"
},
{
"from": "f0119976",
"fromId": "f0119976",
"to": "f01753007",
"toId": "f01753007",
"value": "4559000000000000000000",
"type": "transfer"
}
]

I want to know the information of the actual recipient and the amount of coin sent in the multi-approve message.
Please help me....

**Reply by jazg** [reply]
Hey @LexLuthr, just bumping this — any recommendations for getting the above details for an approve transaction?

**Reply by LexLuthr** [reply]
@jinho-korbit Can you tell me where you got the example output from?

**Reply by i-xiaov** [reply]
Same issue. 
@LexLuthr , it's just internal-transaction, many explorers support it. Usually I use `Filecoin.StateReplay` method to get the `ExecutionTrace`, then I can find out all transfers occurred. But it did not work for an approve transaction, there is no transfer in the execution trace. Any suggestion plz?

**Reply by arajasek** [reply]
@i-xiaov `StateReplay` is definitely the correct method for this, and it should work for `MultisigApprove` messages too. 

Example:

```
 curl -X POST \
                                                  -H "Content-Type: application/json" \
                                                  --data '{ "jsonrpc": "2.0", "method":"Filecoin.StateReplay", "params": [[
                                              {"/":"bafy2bzacebj3kclj7vry3fpblsl5dy42xf4qhedfmlbvr23zp6zfzay53qrc2"}],
                                               {"/": "bafy2bzaced3kvaz3b4fpp453uwtzidjobid32dqmmq4am5f3bd5kifbrjnnka"}], "id": 0 }' 'http://127.0.0.1:1234/rpc/v1'

{"jsonrpc":"2.0","result":{"MsgCid":{"/":"bafy2bzaced3kvaz3b4fpp453uwtzidjobid32dqmmq4am5f3bd5kifbrjnnka"},"Msg":{"Version":0,"To":"f01551055","From":"f1numpwkhmylf3lhdcxulyrautjfjfjucfrqeokhi","Nonce":15414,"Value":"0","GasLimit":37995837,"GasFeeCap":"1481741336","GasPremium":"100321","Method":3,"Params":"ghkCeFggUepbu6a7BfCWNnbU7mVsKhY54inWYdBi0ssxAcj5fTQ=","CID":{"/":"bafy2bzaceaq2qw2537fbsg2ncvlxl2nxj7o6clyqg6uiixvtwyikxofxuxdrq"}},"MsgRct":{"ExitCode":0,"Return":"g/UAS0oAHGnYCjSoGW1l","GasUsed":30420070},"GasCost":{"Message":{"/":"bafy2bzaceaq2qw2537fbsg2ncvlxl2nxj7o6clyqg6uiixvtwyikxofxuxdrq"},"GasUsed":"30420070","BaseFeeBurn":"4395621996260240","OverEstimationBurn":"163149160522560","MinerPenalty":"0","MinerTip":"3811780363677","Refund":"51737419341671755","TotalCost":"4562582937146477"},"ExecutionTrace":{"Msg":{"Version":0,"To":"f01551055","From":"f1numpwkhmylf3lhdcxulyrautjfjfjucfrqeokhi","Nonce":15414,"Value":"0","GasLimit":37995837,"GasFeeCap":"1481741336","GasPremium":"100321","Method":3,"Params":"ghkCeFggUepbu6a7BfCWNnbU7mVsKhY54inWYdBi0ssxAcj5fTQ=","CID":{"/":"bafy2bzaceaq2qw2537fbsg2ncvlxl2nxj7o6clyqg6uiixvtwyikxofxuxdrq"}},"MsgRct":{"ExitCode":0,"Return":"g/UAS0oAHGnYCjSoGW1l","GasUsed":30420070},"Error":"","Duration":531215704,"GasCharges":null,"Subcalls":[{"Msg":{"Version":0,"To":"f025002","From":"f01551055","Nonce":0,"Value":"0","GasLimit":0,"GasFeeCap":"0","GasPremium":"0","Method":16,"Params":"gUoAHGnYCjSoGW1l","CID":{"/":"bafy2bzacecvzt7opd2yr6de3aj6dzydjxttjitwdvmeq6273d6xtjkzv3qplu"}},"MsgRct":{"ExitCode":0,"Return":"SgAcadgKNKgZbWU=","GasUsed":0},"Error":"","Duration":0,"GasCharges":null,"Subcalls":[{"Msg":{"Version":0,"To":"f01551055","From":"f025002","Nonce":0,"Value":"524135691254094196069","GasLimit":0,"GasFeeCap":"0","GasPremium":"0","Method":0,"Params":null,"CID":{"/":"bafy2bzacedb3xfsp5ryrntn4y2pfsqdeezwiv6sy4wh2fb3meogtpvp4prhwm"}},"MsgRct":{"ExitCode":0,"Return":null,"GasUsed":0},"Error":"","Duration":0,"GasCharges":null,"Subcalls":null}]}]},"Error":"","Duration":531215704},"id":0}
```

Can someone share a CID of a message where it isn't working?

**Reply by i-xiaov** [reply]
@arajasek The example you provided looks like correct indeed. But when I call this, I got a null `Subcalls`. I tried with my own node and `https://api.node.glif.io`, both of them responded the same data:
```
{
    "jsonrpc": "2.0",
    "result": {
        "MsgCid": {
            "/": "bafy2bzaced3kvaz3b4fpp453uwtzidjobid32dqmmq4am5f3bd5kifbrjnnka"
        },
        "Msg": {
            "Version": 0,
            "To": "f01551055",
            "From": "f1numpwkhmylf3lhdcxulyrautjfjfjucfrqeokhi",
            "Nonce": 15414,
            "Value": "0",
            "GasLimit": 37995837,
            "GasFeeCap": "1481741336",
            "GasPremium": "100321",
            "Method": 3,
            "Params": "ghkCeFggUepbu6a7BfCWNnbU7mVsKhY54inWYdBi0ssxAcj5fTQ=",
            "CID": {
                "/": "bafy2bzaceaq2qw2537fbsg2ncvlxl2nxj7o6clyqg6uiixvtwyikxofxuxdrq"
            }
        },
        "MsgRct": {
            "ExitCode": 0,
            "Return": "g/UAS0oAHGnYCjSoGW1l",
            "GasUsed": 30420070
        },
        "GasCost": {
            "Message": {
                "/": "bafy2bzaceaq2qw2537fbsg2ncvlxl2nxj7o6clyqg6uiixvtwyikxofxuxdrq"
            },
            "GasUsed": "30420070",
            "BaseFeeBurn": "4395621996260240",
            "OverEstimationBurn": "163149160522560",
            "MinerPenalty": "0",
            "MinerTip": "3811780363677",
            "Refund": "51737419341671755",
            "TotalCost": "4562582937146477"
        },
        "ExecutionTrace": {
            "Msg": {
                "Version": 0,
                "To": "f01551055",
                "From": "f1numpwkhmylf3lhdcxulyrautjfjfjucfrqeokhi",
                "Nonce": 15414,
                "Value": "0",
                "GasLimit": 37995837,
                "GasFeeCap": "1481741336",
                "GasPremium": "100321",
                "Method": 3,
                "Params": "ghkCeFggUepbu6a7BfCWNnbU7mVsKhY54inWYdBi0ssxAcj5fTQ=",
                "CID": {
                    "/": "bafy2bzaceaq2qw2537fbsg2ncvlxl2nxj7o6clyqg6uiixvtwyikxofxuxdrq"
                }
            },
            "MsgRct": {
                "ExitCode": 0,
                "Return": "g/UAS0oAHGnYCjSoGW1l",
                "GasUsed": 30420070
            },
            "Error": "",
            "Duration": 5978340,
            "GasCharges": null,
            "Subcalls": null
        },
        "Error": "",
        "Duration": 5978340
    },
    "id": 0
}
```

**Reply by tok-kkk** [reply]
@i-xiaov According to this #9046, you'll need to set `LOTUS_VM_ENABLE_TRACING` to `true` for your node. 
But I wonder if there's a way to get this info from public nodes (i.e. glif), since I don't run my own node.

**Reply by i-xiaov** [reply]
@tok-kkk Thanks for you replay. It solved my problem perfectly.
Well, I don't think the public nodes will support `StateReplay` method as it's kind of heavy api.
